src/legacy/ui/public/vis/vis_types/vislib_vis_legend/vislib_vis_legend.tsx

```diff
@@ -0,0 +1,239 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import React, { useState, useEffect, BaseSyntheticEvent, KeyboardEvent, memo } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import { compact, uniq, map } from 'lodash';
+
+import { i18n } from '@kbn/i18n';
+import { EuiIcon, keyCodes, htmlIdGenerator } from '@elastic/eui';
+
+// @ts-ignore
+import { Data } from '../../../vislib/lib/data';
+// @ts-ignore
+import { createFiltersFromEvent } from '../../vis_filters';
+import { CUSTOM_LEGEND_VIS_TYPES, LegendItem } from './models';
+import { VisLegendItem } from './vislib_vis_legend_item';
+import { getTableAggs } from '../../../visualize/loader/pipeline_helpers/utilities';
+
+export interface VisLegendComponentProps {
+  vis: any;
+  visData: any;
+  uiState: any;
+  refreshLegend: number;
+}
+
+let getColor: (label: string) => string;
+
+const VisLegendComponent = ({ vis, visData, uiState, refreshLegend }: VisLegendComponentProps) => {
+  const [open, setOpen] = useState(uiState.get('vis.legendOpen', true));
+  const [labels, setLabels] = useState<any[]>([]);
+  const [selectedLabel, setSelectedLabel] = useState<string | null>(null);
+  const [tableAggs, setTableAggs] = useState<any[]>([]);
+  const legendId = htmlIdGenerator()('legend');
+
+  useEffect(() => {
+    if (visData) {
+      refresh();
+    }
+  }, [refreshLegend]);
+
+  const toggleLegend = () => {
+    const bwcAddLegend = vis.params.addLegend;
+    const bwcLegendStateDefault = bwcAddLegend == null ? true : bwcAddLegend;
+    const newOpen = !uiState.get('vis.legendOpen', bwcLegendStateDefault);
+    setOpen(newOpen);
```

Comment:
Can you check whether anything bad happens in we don't even hold the open state locally in the component but just use `vis.legendOpen` directly? I know it's not exactly like the old behavior but this is probably in place because of some weird angular behavior and I would be glad to get rid of it to be able to don't have unnecessary local state in the component.

Reply:
I'll check it out.

Reply:
looks to work fine c913d98